src/common/com/intellij/plugins/haxe/ide/folding/HaxeFoldingBuilder.java

```diff
@@ -58,14 +143,118 @@ private static void buildFolding(@NotNull ASTNode node, List<FoldingDescriptor>
       descriptor = buildCodeBlockFolding(node);
     } else if (isBodyBlock(elementType)) {
       descriptor = buildBodyBlockFolding(node);
+    } else if (isComment(elementType, node)) {
+       RegionMarker matched = matchRegion(node);
+       if (null != matched) {
+         regionMarkers.add(matched);
+       }
+    } else if (isCompilerConditional(elementType)) {
+      RegionMarker matched = matchCCRegion(node);
+      if (null != matched) {
+        ccMarkers.add(matched);
+      }
     }
 
     if (descriptor != null) {
-      list.add(descriptor);
+      descriptors.add(descriptor);
     }
 
     for (ASTNode child : node.getChildren(null)) {
-      buildFolding(child, list);
+      buildFolding(child, descriptors, regionMarkers, ccMarkers);
+    }
+  }
+
+  private static RegionMarker peekUpStack(List<RegionMarker> list, int n) {
+    int current = 0;
+    for (RegionMarker marker : list) {
+      if (current == n) {
+        return marker;
+      }
+      current++;
+    }
+    return null;
+  }
```

Comment:
```java
return n >= 0 && n < list.size() ? list.get(n) : null;
```
:)

Reply:
Yeah, that's what happens when I'm thinking linked list and don't remember that there's a get(n) function.  (And, when I'm tired. :wink: )  I'll fix it shortly...

And, I really appreciate your review.